frost/src/keys.rs

```diff
@@ -0,0 +1,276 @@
+};
+
+/// The public key used to verify a threshold signature made by a group of signers.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct GroupPublicKey<G: AffineCurve>(pub G);
+
+/// The signer's public key.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SignerPublicKey<G: AffineCurve>(pub G);
+
+/// The signer's secret key
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SignerSecretKey<G: AffineCurve>(pub G::ScalarField);
+
+/// The list of signer public keys and the group public key.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PublicKeys<G: AffineCurve> {
+    /// The map of all participant public keys.
+    pub public_keys: HashMap<u64, SignerPublicKey<G>>,
+    /// The group public key used to verify the final threshold signature.
+    pub group_public_key: GroupPublicKey<G>,
+}
+
+/// A signer's share that includes its secret key and all publicly known keys/commitments.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SignerShare<G: AffineCurve> {
+    /// The index of the participant.
+    pub participant_index: u64,
+    /// The participant's secret key share.
+    pub secret_key: SignerSecretKey<G>,
+    /// The participant's public key.
+    pub public_key: SignerPublicKey<G>,
+    /// The group public key used to verify the final threshold signature.
+    pub group_public_key: GroupPublicKey<G>,
+    /// The public commitments to the coefficients (generated by each participant).
+    pub commitment: Vec<G>,
+}
+
+impl<G: AffineCurve> SignerShare<G> {
+    /// Verify that the secret share was generated correctly and matches the commitment.
+    pub fn is_valid(&self) -> bool {
+        let expected_result = G::prime_subgroup_generator().mul(self.secret_key.0);
+
+        let index_scalar = G::ScalarField::from_repr((self.participant_index as u64).into()).unwrap();
+        let mut result = G::zero().to_projective();
+        for (i, c) in self.commitment.iter().rev().enumerate() {
+            result = result.add(c.to_projective());
+
+            if i != self.commitment.len() - 1 {
+                result = result.mul(index_scalar);
+            }
+        }
+
+        expected_result == result.into()
+    }
+}
+
+/// Generate the participant keys using a trusted authority/dealer. This process can be done with
+/// DKG, but for simplicity we will use a trusted dealer model. This trusted dealer model only
+/// generates a single polynomial for all participants, and then uses the indexes to determine
+/// secret shares for each participant.
+///
+/// The `secret` attribute is randomly sampled and passed through for testing purposes.
+/// g^`secret` = GroupPublicKey
+pub fn trusted_keygen<R: Rng, G: AffineCurve>(
+    num_participants: u8,
+    threshold: u8,
+    secret: &G::ScalarField,
+    rng: &mut R,
+) -> (Vec<SignerShare<G>>, PublicKeys<G>) {
+    if num_participants < 1 {
+        panic!("The number of participants must be greater than 0.");
+    }
+
+    if threshold < 1 || threshold > num_participants {
```

Comment:
```suggestion
    if threshold == 0 || threshold > num_participants {
```
`clippy` is likely to suggest this

Reply:
the same applies to line 90